src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/Utf8BufferTextReader.cs

```diff
@@ -0,0 +1,54 @@
+﻿using System;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Microsoft.AspNetCore.SignalR.Internal.Protocol
+{
+    internal class Utf8BufferTextReader : TextReader
+    {
+        private ReadOnlyMemory<byte> _utf8Buffer;
+
+        public Utf8BufferTextReader(ReadOnlyMemory<byte> utf8Buffer)
```

Comment:
Every time I read `HubEndPoint` and see the `ToArray` line 150 I don't like it. See below.

What about if `HubProtocol` took a sequence, which was then passed to this text reader. The logic in `Utf8BufferTextReader` would stay largely the same expect it could read over each memory block in the sequence instead of just one. We can avoid copying the sequence in its entirety into a new array.

https://github.com/aspnet/SignalR/blob/8b8c7107a8e0b4a2123f8a8178525f1da8aaeb16/src/Microsoft.AspNetCore.SignalR.Core/HubEndPoint.cs#L138-L168

Reply:
This change will be made after @anurse 's refactoring. The problem is that right now the client and server are asymmetrical and they don't both parse `ReadOnlySequence<byte>`. Client uses a byte[] and the server uses `ReadOnlySequence<byte>`.  This will be done as part of https://github.com/aspnet/SignalR/issues/1508.

Reply:
Could we make the server and `Utf8BufferTextReader` work with `ReadOnlySequence<byte>` now? On the client we can just wrap the `byte[]` in a `ReadOnlySequence<byte>` with little to no overhead before parsing, no?

Reply:
I don't really see the point.

Reply:
I also hate ToArray() with a passion but there is a master plan here and it's going to be glorious when it comes together.